system/lib/wasmfs/wasmfs.cpp

```diff
@@ -67,9 +75,14 @@ WasmFS::~WasmFS() {
 }
 
 std::shared_ptr<Directory> WasmFS::initRootDirectory() {
+  
+#ifdef WASMFS_CASE_INSENSITIVE
+  auto rootBackend = wasmfs_create_icase_backend(make_mem_backend, nullptr);
+#else
   auto rootBackend = createMemoryFileBackend();
-  auto rootDirectory =
-    std::make_shared<MemoryDirectory>(S_IRUGO | S_IXUGO | S_IWUGO, rootBackend);
+#endif
+  auto rootDirectory = reinterpret_cast<wasmfs::Backend*>(rootBackend)
```

Comment:
There shouldn't be any need for the `reinterpret_cast` here.

Reply:
We need it because the declaration of `wasmfs_create_icase_backend()` refers to `backend_t` which are both under `extern "C"`:

```
wasmfs.cpp:76:35: error: member access into incomplete type 'struct Backend'
  auto rootDirectory = rootBackend->createDirectory(S_IRUGO | S_IXUGO | S_IWUGO);
                                  ^
emscripten/cache/sysroot/include/emscripten/wasmfs.h:17:16: note: forward declaration of 'Backend'
typedef struct Backend* backend_t;
               ^
1 error generated.
```
This is not the case for `createMemoryFileBackend()`.
I can add another declaration of a factory function to avoid reinterpret cast.

Reply:
Oh I see 👍

Reply:
Maybe we can add a simpler C++ interface for this (like `createMemoryFileBackend()` for the in-memory backend) so we don't need to use the more complex C API here?

@kripken, wdyt about this? I know we had discussed not having a separate case-insensitive version of libwasmfs and not supporting a case-insensitive root directory going forward, but this seems fairly simple and avoids regressing functionality that we currently support.

Reply:
C++ APIs for backend creation that have more type info sgtm.